client/src/editor/editor-context.tsx

```diff
@@ -7,20 +7,25 @@ export interface EditorState {
     processingFilter: BaseFilter | null;
     applyFilter: (filter: BaseFilter, options: any) => Promise<void>;
     imageHistory: ImageHistory;
+    updateImageHistory: (updatedHistory: ImageHistory) => void;
 }
 
 const defaultState: EditorState = {
     processingFilter: undefined,
     applyFilter: undefined,
     imageHistory: undefined,
+    updateImageHistory: undefined // Add this line
+
 };
 
 export const EditorContext = createContext<EditorState>(defaultState);
 
 export const EditorContextProvider: FC<{ children: ReactNode }> = ({ children }) => {
     const { imageData } = useContext(AppContext);
     const [processingFilter, setProcessingFilter] = useState(null);
-    const imageHistory = useMemo(() => new ImageHistory([imageData]), [imageData]);
+    const [imageHistory, setImageHistory] = useState(() => new ImageHistory([imageData]));
+
+
```

Comment:
it's not good to update image history for every operation. I think it's a bad design. Can you try to this way: create a property index in image history that point to the current image and find a way to re-render components when the value of index changes